sql: general improvements in the handling of qualified names

knz opened the pull request:
I am in the process of prototyping the proposal in #21456 and on the way to a prototype I discovered both a major need to clean up the code and a very nice opportunity to do so, so I briefly shaved that yak.

The benefit is a 10-20% improvement on parsing times and memory costs. Will be most visible in complex statements with numerous column references.

```
name                              old time/op    new time/op    delta
FormatRandomStatements/parse-16     4.81ms ±57%    4.01ms ± 5%  -16.70%  (p=0.009 n=10+8)
FormatRandomStatements/format-16     630µs ± 3%     580µs ± 0%   -7.96%  (p=0.000 n=8+9)

name                              old alloc/op   new alloc/op   delta
FormatRandomStatements/parse-16     1.72MB ± 0%    1.64MB ± 0%   -4.79%  (p=0.000 n=10+8)
FormatRandomStatements/format-16    58.6kB ± 0%    58.9kB ± 0%   +0.44%  (p=0.000 n=10+10)

name                              old allocs/op  new allocs/op  delta
FormatRandomStatements/parse-16      26.2k ± 0%     22.8k ± 0%  -13.05%  (p=0.000 n=10+10)
FormatRandomStatements/format-16     1.10k ± 0%     1.09k ± 0%   -1.36%  (p=0.000 n=10+10)
```

knz committed:
sql, sem/tree: rename the 'Database' field to 'Schema'

We're going to address the fact that the current syntactic field used
to determine the database corresponds to the schema in PostgreSQL.

The first step to enlightenment is to realize the extent of the
damage.  This commit does just this: it renames that syntactic
position in the AST types "TableName" and "AllTablesSelector" to
"Schema" and propagates the change.

Every where else where a "database name" is being assigned to /
compared with a field called "Schema" should then be considered wrong
and in want of correction. As far as I can see there is only one
location (extracting the virtual table descriptor inside a virtual
schema) which is doing the right thing.

Subsequent commits will improve the situation incrementally.

Release note: None

knz committed:
sql, sql/parser, sql/sem/tree: simplify the handling of names

Prior to this patch, a single non-terminal `qualified_name` in the
grammar was handling all the other grammar needs for naming things
with a path that optionally includes a star (x.y.z), `any_name` for a
path that cannot include a star, and `name` and `unrestricted_name`
were sprinkled around the other rules for things without a path,
without apparent structure.

In a future patch we want to properly discriminate between names that
designate a database object (with a possible db and schema prefix)
from those names that designate other things, so as to only apply
the schema vs catalog normalization rules on the former category.

As a prerequisite to this change, we want the grammar to properly
reveal which name belongs to which category.

This patch does just this, by using different non-terminals in the
grammar for different kinds of names:

- `table_name` for a potentially-qualified table name with no star
- `table_pattern` for a potentially-qualified table name or glob
- `column_name` for a single unqualified column name
- `column_path` for a potentially-qualified column name with no star
- `column_path_with_star` for a potentially-qualified column name or glob
- `var_name` for a variable name potentially containing periods

etc.

In addition, it avoids a parsing recursion for `table_name`,
`column_path` variants and `table_pattern` and reveals the various
fixed formats that the various names can have. In various places,
starting with function names, a star is now rejected during parsing
as a syntax error instead of during name resolution (= less work).

While authoring this patch, two side effects were obtained:

- the handling of INSERT, UPDATE and UPSERT was simplified, by
  avoiding paying the cost of name normalization for the target column
  list.  Since CockroachDB currently only supports simple names in
  this position, the grammar is simplified accordingly.
- a bug in the grammar accepted for COPY was fixed: the grammar
  previously recognized the syntax "`COPY foo() FROM STDIN`", whereas
  an empty column list between the parentheses is incorrect according
  to pg's grammar.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21753)
<!-- Reviewable:end -->

knz committed:
sql/sem/tree: make IMPORT use normalizable table references

... for consistency with every other use of `table_name`.

Release note: None

knz committed:
sql, sql/parser, sql/sem/tree: sanitize/simplify the handling of names

Prior to this patch, the type `UnresolvedName` was an array of things
called "name parts", and could include stars, array indirections or
subscripts. The corresponding logic to normalize unresolved names into
table names, column names or other things was complicated.

In the light of a desired upcoming change to support both catalog and
schema specifiations in names, this code was welcoming a
simplification.

The simplification came from the realization that the syntactic
structure for a table or column path is unambiguous and the "item of
interest" is always in the last position. My previous (mistaken)
understanding was that a column name could be followed by other name
parts in the grammar; however, this is not valid SQL.

- to access an array column, the grammar rules to parse `a.b.c[d]`
  parse first `a.b.c` as a column item, then `<expr> [d]` as an array
  subscript expression. The `[d]` part is not part of the grammar for
  names, never has.

- to access a field in a column of a compound type (these are not
  supported by CockroachDB yet, but the grammar mirrors pg's which
  does support them), pg's SQL dialect has it that one *must* write
  `(a.b.c).field` with grouping parentheses; a direct path of
  `a.b.c.field` is not valid. (More precisely, `a.b.c.field`
  *always* refers to the column `field` of table `c` in schema `b` in
  catalog `a`, and the query would error out if there is no such
  table/column.)

  In short, an unresolved column name again always has a period-separated
  path with the column name in last position.

The predecessor to this patch already had simplified the grammar rules
to account for the known maximum length of a qualified name. This
patch now simplifies the AST and all uses to further capitalize on the
new understanding and simplify the normalization and pretty-printing
rules.

Release note: None

knz commented:
Also just to be sure I did run all the benchmarks in `pkg/bench` and `benchstat` picked up no significant difference.

knz committed:
sql/sem/tree, sql: clarify TableName.SchemaName and TableName.CatalogName

Working slowly towards a better use of the catalog prefix in `TableName`,
this patch clarifies the struct:

- the "Prefix" field is renamed to "Catalog".
- the fields "OmitSchemaDuringFormatting" and
  "PrefixOriginallySpecified" are renamed to "ExplicitSchema" (meaning
  inverted) and "ExplicitCatalog", to clarify what they are really
  about.
- new helper methods are provided to construct instances of
  `TableName`. This sensibly shortens most points of use, and
  clarifies intent.

Release note: None

petermattis commented:
I thought we took the name portion of the grammar directly from the Postgres grammar. You're claiming that the Postgres grammar is different from ours. I just want to verify that this PR is moving us closer to the Postgres approach.

---

Review status: 0 of 99 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-:-L3mxWwhC7maSnqOcpB6:bvawxsb)*
<!-- Sent from Reviewable.io -->

knz commented:
Postgres contains two additional things in their grammar which CockroachDB does not support, neither before or with this patch:

- inserting or updating fields in compound types: 
  - `insert into foo (a[123], b.c, b.c.d[123].e.f) values ...`
  - `update foo set a[123] = 456, (b.c, b.c.d[123].e.f) = (1, 2)...`
- accessing a field in a column item, e.g. `select (a).b.c from xyz` (this says "access field c of compound item b that is itself a field of column a". The notation `a.b.c` is *not* suitable to achieve the same, neither in pg nor in cockroachdb, as explained by the commit messages)

In CockroachDB with this patch, we need later to:
- extend the rule for `insert_column_item` to support subscripting of the insertion/update targets.
- add new rules to support expressions of the form `( <expr> ).a.b.c`

There are two salient remarks about pg's implementaiton:
- in pg these were supported from very early on and some effort was made to *share grammar rules* across these various features.  This sharing is only meaningful when the support is symmetric. In CockroachDB the support is not symmetric so the sharing of grammar rules is actually creating more complexity (because we need to exclude some unsupported constructs after parsing)
- pg handles all these things, *including the path to a table or column name* using extremely lightweight linked lists in memory with a sane/eager/exact memory reclamation strategy. Therefore, the choice of a recursive rule was a good match. In Go, we absolutely avoid linked lists, and we try to avoid allocating slices on the heap, when a fixed-sized data structure is sufficient. In this patch, I found out that these "lists" are actually arrays of a very well-specified maximum size (4), so the overhead of a "linked list treatment" in the grammar was really not justified.

jordanlewis commented on pkg/sql/parser/sql.y:
```diff
@@ -6206,14 +6177,11 @@ c_expr:
 
 // Productions that can be followed by a postfix operator.
 //
```
I don't quite understand why you needed to introduce all of these new names. Was it not enough to just leave the RHS of these in the various rules?

jordanlewis commented on pkg/sql/parser/sql.y:
```diff
@@ -6260,27 +6228,27 @@ d_expr:
-    $$.val = &tree.FuncExpr{Func: $1.resolvableFunctionReferenceFromUnresolvedName(), Type: tree.DistinctFuncType, Exprs: $4.exprs()}
+    $$.val = &tree.FuncExpr{Func: $1.resolvableFuncRefFromName(), Type: tree.DistinctFuncType, Exprs: $4.exprs()}
   }
 | func_name '(' '*' ')'
```
How will we extend this for compound types? Could you put a TODO here since you've got your mind in this area right now?

jordanlewis commented on pkg/sql/parser/sql.y:
```diff
@@ -6483,7 +6451,7 @@ over_clause:
     $$.val = $2.windowDef()
   }
-| OVER name
+| OVER window_name
```
Is this part tested? I guess we'd need to test `db.public.table.col`, right?

jordanlewis commented on pkg/sql/parser/sql.y:
```diff
@@ -6862,38 +6830,6 @@ opt_slice_bound:
-  {
-    $$.val = tree.UnqualifiedStar{}
-  }
-
```
This is very nice documentation.

jordanlewis commented on pkg/sql/sem/tree/import.go:
```diff
@@ -16,7 +16,7 @@ package tree
 // Import represents a IMPORT statement.
 type Import struct {
-	Table      UnresolvedName
+	Table      NormalizableTableName
```
I'm not completely sure what effect this has on IMPORT, but ISTM you need a new test to prove it one way or another, right?

jordanlewis commented on pkg/sql/sem/tree/eval.go:
```diff
@@ -2872,9 +2872,13 @@ func performCast(ctx *EvalContext, d Datum, t coltypes.CastTargetType) (Datum, e
-					name = append(name, (*Name)(&substrs[i]))
+				if len(substrs) > 3 {
+					return nil, pgerror.NewErrorf(pgerror.CodeSyntaxError,
+						"invalid function name: %s", s)
```
What is this case and is it tested?

jordanlewis commented on pkg/sql/sem/tree/function_definition.go:
```diff
@@ -67,17 +67,15 @@ func (fd *FunctionDefinition) String() string { return AsString(fd) }
+		// n.NumParts > 2 would be catalog.schema.funname().
+		// We don't support those. (yet?)
+		return nil, pgerror.NewErrorf(pgerror.CodeSyntaxError,
+			"invalid function name: %q", ErrString(n))
```
We will need this soon I think. Another thing we need while I'm thinking about it is namespaced types like `::pg_catalog.regclass`.

jordanlewis commented on pkg/sql/sem/tree/table_name.go:
```diff
@@ -94,33 +94,73 @@ type TableNameReference interface {
+	tn := MakeTableName(db, tbl)
+	return &tn
+}
+
```
This form is weird. Can we reconsider it now that we're switching the semantics of our naming hierarchy? Remind me why we need it again?

jordanlewis commented on pkg/sql/set_var.go:
```diff
@@ -39,13 +40,13 @@ type setVarNode struct {
+	if n.Name == "" {
+		// A client has sent the reserved internal syntax SET ROW ...,
+		// or the user entered `SET "" = foo`. Reject it.
+		return nil, pgerror.NewErrorf(pgerror.CodeInvalidNameError, "invalid variable name: %q", n.Name)
```
Is it tested?

jordanlewis reviewed (commented):
Generally LGTM! Nice insight and resultant performance improvement.
<!-- review by re -->

jordanlewis commented:
My review tool submitted horribly mis-aligned comments in `sql.y` for some reason. The other comments are located correctly. I will correct this in a moment.

petermattis commented:
@knz Thanks for the clarification.

knz commented:
Review status: 0 of 103 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/set_var.go, line 46 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-L3hnzeVS2P3VXFz-eWT-r5-46:-L3oh0CF98lfVffUEMJ1:b-c7baj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e7e2bf39aa06a3ba249197c90d5936dfe02033/pkg/sql/set_var.go#L46)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Is it tested?
</blockquote></details>

Yes `logic_tests/set` - SET ROW.

---

*[pkg/sql/parser/sql.y, line 6179 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-L3dBdcBqdeQhdXeikr8-r5-6179:-L3ofbx9WrIWfx8pltb6:b-ud6k9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e7e2bf39aa06a3ba249197c90d5936dfe02033/pkg/sql/parser/sql.y#L6179)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I don't quite understand why you needed to introduce all of these new names. Was it not enough to just leave the RHS of these in the various rules?
</blockquote></details>

Added a comment to explain/motivate.

---

*[pkg/sql/parser/sql.y, line 6247 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-L3dBdcBqdeQhdXeikr8-r5-6247:-L3ofectfP4sSkK0SUlm:b-80wdnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e7e2bf39aa06a3ba249197c90d5936dfe02033/pkg/sql/parser/sql.y#L6247)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

How will we extend this for compound types? Could you put a TODO here since you've got your mind in this area right now?
</blockquote></details>

Added several todo's / comments in the right places in the grammar to explain how to do compound types.

---

*[pkg/sql/parser/sql.y, line 6454 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-L3dBdcBqdeQhdXeikr8-r5-6454:-L3ofngix99bVQHhug6i:bbjtgjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e7e2bf39aa06a3ba249197c90d5936dfe02033/pkg/sql/parser/sql.y#L6454)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Is this part tested? I guess we'd need to test `db.public.table.col`, right?
</blockquote></details>

Added a logic test to verify that the extended form is rejected.

---

*[pkg/sql/sem/tree/eval.go, line 2877 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-L3dBdcBqdeQhdXeikqr-r5-2877:-L3oiGbshiFMiJx_ddx0:b-ie8mhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e7e2bf39aa06a3ba249197c90d5936dfe02033/pkg/sql/sem/tree/eval.go#L2877)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What is this case and is it tested?
</blockquote></details>

Added a comment here and a logic test in `logic_tests/pgoidtype`.

---

*[pkg/sql/sem/tree/function_definition.go, line 74 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-L3hnzeVS2P3VXFz-eWP-r5-74:-L3oiPLaILl5ohd1MhHh:b-aajso2) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e7e2bf39aa06a3ba249197c90d5936dfe02033/pkg/sql/sem/tree/function_definition.go#L74)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

We will need this soon I think. Another thing we need while I'm thinking about it is namespaced types like `::pg_catalog.regclass`.
</blockquote></details>

Namespaced types will be a different work item.
The 3-part function name will come with the code I'm preparing for #21456.

---

*[pkg/sql/sem/tree/import.go, line 19 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-L3gopTBM6Q8Zpiz6gWu-r5-19:-L3ogkk7zmfm-UgHr5Pa:b2cr7z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e7e2bf39aa06a3ba249197c90d5936dfe02033/pkg/sql/sem/tree/import.go#L19)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'm not completely sure what effect this has on IMPORT, but ISTM you need a new test to prove it one way or another, right?
</blockquote></details>

There is no change in semantics, it's just DRY on code.
The ccl package was re-implementing explicitly what NormalizableTableName already does. I deleted the code in there as a result of introducing NormalizableTableName here.

---

*[pkg/sql/sem/tree/table_name.go, line 129 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-L3dBdcBqdeQhdXeikqt-r5-129:-L3oiaoGpqjHSLEiyJ2k:bop7qep) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e7e2bf39aa06a3ba249197c90d5936dfe02033/pkg/sql/sem/tree/table_name.go#L129)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This form is weird. Can we reconsider it now that we're switching the semantics of our naming hierarchy? Remind me why we need it again?
</blockquote></details>

We need it for our already-supported access to virtual tables: `select * from mydb.information_schema.tables` restricts the list of tables visible in the vtable to those of `mydb`. We created that code to avoid spilling all the cluster's tables from other dbs from client connected to 1 db.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Reviewable cannot handle this PR. I'm going to diff your two patchsets by hand. Argh

jordanlewis commented:
Ok I can't really figure out how to effectively review this. It <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but for the record it's very hard to review this size of PR. I would prefer in the future that you make your PRs have fewer commits. For example, it would have been much easier to review this if you had the first PR *only* rename the fields, etc.

---

Review status: 0 of 103 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21753#-:-L41R6etRlPVT2lx4d3u:bk5d0ud)*
<!-- Sent from Reviewable.io -->

knz commented:
Closing in favor of #22266, #22267, #22268, #22269, #22270 to facilitate the review.
